lib/AST/ASTPrinter.cpp

```diff
@@ -907,14 +907,11 @@ class PrintAST : public ASTVisitor<PrintAST> {
         DC = DC->getParent()->getInnermostTypeContext();
```

Comment:
Is it possible that this code is where the DC is computed incorrectly?

If 'Current' is a NominalTypeDecl, DC becomes Current->getInnermostDeclContext(), which is a child of the one you want.

I'm guessing for the case where Current is an AbstractFunctionDecl, you really do want the innermost DC for the mapTypeOutOfContext() call to succeed.

Try replacing

    DC = DC->getInnermostTypeContext();

with

    if (isa<AbstractFunctionDecl>(Current) || isa<GenericTypeDecl>(Current)) DC = DC->getParent();

then remove the TypeAliasDecl special case too.

this way we still skip up a level if necessary, even if DC already is a type context (which it will be for a reference to a nested type).